### ISSUE TITLE
Getting false positives for 'Alt' key.

### ISSUE THREAD

**Issue by leMicin** [TolikPylypchuk/SharpHook#59]
We have a piece of software that tries to listen to certain key combinations. When we use the 'Alt' key in version 5.x, it sometimes gets stuck in the 'pressed' state and stay that way until we manually press it again. This issue didn't happen with version 4.2.x. 

``` cs
        private void OnKeyPressed(object? sender, KeyboardHookEventArgs args)
        {
            // [...]
            if ((args.RawEvent.Mask & ModifierMask.Alt) > 0)
            {
                // this gets here even when alt is not pressed down.
            }
```

Our full implementation of the class is [here](https://github.com/Sidekick-Poe/Sidekick/blob/feature/wpf/src/Sidekick.Common.Platform/Keyboards/KeyboardProvider.cs).

I have tried to comment out pieces of code to get to the bare minimum and the false positives appear to stay there. Then when I 'spam' the alt key to kind of unstick it, keybinds work as expected.

**Comment by TolikPylypchuk** [comment]
Hi! Thanks for posting this issue! Could you please provide a little more info? I need to know which OS it is, for example. Also, it would be great if you shared [libuiohook debug logs](https://sharphook.tolik.io/v5.0.0/articles/logging.html) so that I can get a better idea of what's going on in libuiohook when you use the global hook (I'm almost sure it's a problem in libuiohook itself).

**Comment by TolikPylypchuk** [comment]
Unrelated thing, but I've noticed that you're suppressing the event inside the event handler. This doesn't work with `TaskPoolGlobalHook` since event suppression must be synchronous, i.e. on the same thread as the event handler of libuihook. You should instead use `SimpleGlobalHook` - it runs its event handlers on the same thread, but make sure that the event handler doesn't run for a long time, or the OS won't like it (I see that the event handler is short, so it should be OK). Also, if your app is cross-platform, then bear in mind that event suppression works only on Windows and macOS.

**Comment by leMicin** [comment]
Thanks for taking the time to look into this. I will get you the logs when I next see the problem.

We are currently only supporting Windows. We had a goal of cross platform at some point, but that goal keeps moving as we always struggle to find cross platform solutions.

**Comment by leMicin** [comment]
### Log implementation
I implemented the logging as you described. The minimum level is debug as anything less would give no log information. I had to adjust the number of functions that were written to logs. Let me know if some of these functions are necessary for you to debug. The relevant pieces of the implementation are as follows:

``` cs
            // Configure hook logging
            LogSource = LogSource.Register(minLevel: SharpHook.Native.LogLevel.Debug);
            LogSource.MessageLogged += OnMessageLogged;
```
``` cs
        private Regex IgnoreHookLogs = new Regex("(?:dispatch_mouse_move|hook_get_multi_click_time|dispatch_event|win_hook_event_proc|dispatch_mouse_wheel|dispatch_button_press|dispatch_button_release)", RegexOptions.Compiled);

        private void OnMessageLogged(object? sender, LogEventArgs e)
        {
            if (IgnoreHookLogs.IsMatch(e.LogEntry.Function))
            {
                return;
            }

            switch (e.LogEntry.Level)
            {
                case SharpHook.Native.LogLevel.Debug:
                    logger.LogDebug("[KeyboardHook] {0}", e.LogEntry.FullText);
                    break;

                case SharpHook.Native.LogLevel.Info:
                    logger.LogInformation("[KeyboardHook] {0}", e.LogEntry.FullText);
                    break;

                case SharpHook.Native.LogLevel.Warn:
                    logger.LogWarning("[KeyboardHook] {0}", e.LogEntry.FullText);
                    break;

                case SharpHook.Native.LogLevel.Error:
                    logger.LogError("[KeyboardHook] {0}", e.LogEntry.FullText);
                    break;
            }
        }
```

### Working keybind (F6)
In the game, press F6 to open a window.
```
2023-08-29 17:41:29.379 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X75 pressed. (0X75)
2023-08-29 17:41:29.382 -04:00 [DBG] [Keyboard] Received key pressed event F6.
2023-08-29 17:41:29.382 -04:00 [DBG] [Keyboard] Executing keybind handler for F6.
2023-08-29 17:41:29.382 -04:00 [DBG] [KeyboardHook] keyboard_hook_event_proc [117]: Consuming the current event. (-1)
2023-08-29 17:41:29.447 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X75 released. (0X75)
2023-08-29 17:41:29.587 -04:00 [DBG] [Keyboard] Completed Keybind Handler for F6.
2023-08-29 17:41:29.719 -04:00 [DBG] Adding entry for "5597b493-f106-4955-8b0a-7bd9c97f84a2"/MudBlazor.MudPopoverProvider. 1 total observers after add.
2023-08-29 17:41:29.719 -04:00 [DBG] Updating entry for "5597b493-f106-4955-8b0a-7bd9c97f84a2"/MudBlazor.MudPopoverProvider. 1 total observers.
2023-08-29 17:41:29.932 -04:00 [DBG] Adding entry for MudBlazor.BrowserViewportSubscription/MudBlazor.MudDrawer. 1 total observers after add.
```

### What triggers the modifier key to stay 'pressed'
In Visual Studio or any text editor, I press Alt+Down.
```
2023-08-29 17:41:39.751 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
2023-08-29 17:41:39.752 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:40.017 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X28 pressed. (0X28)
2023-08-29 17:41:40.018 -04:00 [DBG] [Keyboard] Received key pressed event Alt+Down.
2023-08-29 17:41:40.018 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:40.091 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X28 released. (0X28)
2023-08-29 17:41:40.194 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0XA012 released. (0XA4)
```

### Log of the failing keybind. We now receive Alt+F6 instead of only F6
In the game, press F6 to open a window. No window opens as it detects Alt+F6
```
2023-08-29 17:41:44.042 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X75 pressed. (0X75)
2023-08-29 17:41:44.043 -04:00 [DBG] [Keyboard] Received key pressed event Alt+F6.
2023-08-29 17:41:44.043 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:44.125 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X75 released. (0X75)
```

### Full Logs
```
2023-08-29 17:41:18.412 -04:00 [DBG] Adding entry for "4efb3cd4-0490-4286-b196-fffffe0e856e"/MudBlazor.MudPopoverProvider. 1 total observers after add.
2023-08-29 17:41:18.424 -04:00 [DBG] Updating entry for "4efb3cd4-0490-4286-b196-fffffe0e856e"/MudBlazor.MudPopoverProvider. 1 total observers.
2023-08-29 17:41:18.537 -04:00 [INF] [Initiazation] Initializing Sidekick.Common.Localization.UILanguageProvider
2023-08-29 17:41:18.641 -04:00 [INF] [Initiazation] Initializing Sidekick.Common.Game.Languages.GameLanguageProvider
2023-08-29 17:41:18.755 -04:00 [INF] [Initiazation] Initializing Sidekick.Common.Platform.Windows.Processes.ProcessProvider
2023-08-29 17:41:18.858 -04:00 [INF] [Initiazation] Initializing Sidekick.Common.Platform.Keyboards.KeyboardProvider
2023-08-29 17:41:18.890 -04:00 [DBG] [KeyboardHook] hook_set_dispatch_proc [41]: Setting new dispatch callback to 0X00007FF7D4543DA4.
2023-08-29 17:41:18.896 -04:00 [DBG] [KeyboardHook] hook_run [413]: SetWindowsHookEx() successful.
2023-08-29 17:41:18.896 -04:00 [DBG] [KeyboardHook] refresh_locale_list [369]: GetKeyboardLayoutList(0, NULL) found 3 layouts.
2023-08-29 17:41:18.896 -04:00 [DBG] [KeyboardHook] refresh_locale_list [384]: Received 3 locales.
2023-08-29 17:41:18.896 -04:00 [DBG] [KeyboardHook] get_keyboard_layout_file [325]: Found keyboard layout "00000409".
2023-08-29 17:41:18.896 -04:00 [DBG] [KeyboardHook] refresh_locale_list [458]: Loading layout for 0X0000000004091009: KBDUS.DLL.
2023-08-29 17:41:18.897 -04:00 [DBG] [KeyboardHook] get_keyboard_layout_file [325]: Found keyboard layout "00000409".
2023-08-29 17:41:18.897 -04:00 [DBG] [KeyboardHook] refresh_locale_list [458]: Loading layout for 0X0000000004090409: KBDUS.DLL.
2023-08-29 17:41:18.897 -04:00 [DBG] [KeyboardHook] get_keyboard_layout_file [325]: Found keyboard layout "00001009".
2023-08-29 17:41:18.897 -04:00 [DBG] [KeyboardHook] refresh_locale_list [458]: Loading layout for 0X0000000010090C0C: KBDCA.DLL.
2023-08-29 17:41:18.898 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0XA011 released. (0XA2)
2023-08-29 17:41:18.899 -04:00 [DBG] [KeyboardHook] load_input_helper [623]: refresh_locale_list() found 3 locale(s).
2023-08-29 17:41:18.982 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Parser.Patterns.ParserPatterns
2023-08-29 17:41:19.088 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Metadatas.ItemMetadataProvider
2023-08-29 17:41:19.219 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Static.ItemStaticDataProvider
2023-08-29 17:41:19.328 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Modifiers.EnglishModifierProvider
2023-08-29 17:41:19.457 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Modifiers.ModifierProvider
2023-08-29 17:41:19.782 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.Poe.Pseudo.PseudoModifierProvider
2023-08-29 17:41:19.783 -04:00 [INF] Pseudo stat service initialization started.
2023-08-29 17:41:19.837 -04:00 [INF] 13 attributes fetched.
2023-08-29 17:41:20.037 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.PoeNinja.PoeNinjaClient
2023-08-29 17:41:20.138 -04:00 [INF] [Initiazation] Initializing Sidekick.Apis.PoeWiki.PoeWikiDataProvider
2023-08-29 17:41:24.498 -04:00 [ERR] Unable to find view https://0.0.0.0/
2023-08-29 17:41:29.379 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X75 pressed. (0X75)
2023-08-29 17:41:29.382 -04:00 [DBG] [Keyboard] Received key pressed event F6.
2023-08-29 17:41:29.382 -04:00 [DBG] [Keyboard] Executing keybind handler for F6.
2023-08-29 17:41:29.382 -04:00 [DBG] [KeyboardHook] keyboard_hook_event_proc [117]: Consuming the current event. (-1)
2023-08-29 17:41:29.447 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X75 released. (0X75)
2023-08-29 17:41:29.587 -04:00 [DBG] [Keyboard] Completed Keybind Handler for F6.
2023-08-29 17:41:29.719 -04:00 [DBG] Adding entry for "5597b493-f106-4955-8b0a-7bd9c97f84a2"/MudBlazor.MudPopoverProvider. 1 total observers after add.
2023-08-29 17:41:29.719 -04:00 [DBG] Updating entry for "5597b493-f106-4955-8b0a-7bd9c97f84a2"/MudBlazor.MudPopoverProvider. 1 total observers.
2023-08-29 17:41:29.932 -04:00 [DBG] Adding entry for MudBlazor.BrowserViewportSubscription/MudBlazor.MudDrawer. 1 total observers after add.
2023-08-29 17:41:39.751 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
2023-08-29 17:41:39.752 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:40.017 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X28 pressed. (0X28)
2023-08-29 17:41:40.018 -04:00 [DBG] [Keyboard] Received key pressed event Alt+Down.
2023-08-29 17:41:40.018 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:40.091 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X28 released. (0X28)
2023-08-29 17:41:40.194 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0XA012 released. (0XA4)
2023-08-29 17:41:44.042 -04:00 [DBG] [KeyboardHook] dispatch_key_press [165]: Key 0X75 pressed. (0X75)
2023-08-29 17:41:44.043 -04:00 [DBG] [Keyboard] Received key pressed event Alt+F6.
2023-08-29 17:41:44.043 -04:00 [DBG] [KeyboardHook] keycode_to_unicode [590]: Using keyboard layout 0X0000000004091009.
2023-08-29 17:41:44.125 -04:00 [DBG] [KeyboardHook] dispatch_key_release [237]: Key 0X75 released. (0X75)
```

**Comment by leMicin** [comment]
Also of note, I have seen other cases with other modifier keys such as Shift and Ctrl. I have a harder time making the handler fail as consistently as with the Alt key.

**Comment by TolikPylypchuk** [comment]
It appears that this issue is caused by libuiohook not receiving key up events (as described in issue #61). On Windows libuiohook manages the modifier mask by itself, simply adding/removing current modifiers on key down/up events. This is why if it doesn't receive the key up event for Alt, it will think that it's still pressed and will add it to the modifier masks of future events.

I could reproduce this issue (typing really fast and seeing that keys sometimes remained in the pressed state), but I don't know why that happens and how to fix that. Googling doesn't help either. I think this is a problem with Windows itself, but not sure.

So, I don't really know what to do about this issue. I'll try googling some more and trying to debug it further, but no promises about the result.

**Comment by leMicin** [comment]
Thanks for looking into this.

I mentioned it in the initial post, but this issue doesn't happen with version 4. It started appearing with version 5. We can keep using version 4 for now with no issue.

I understand this kind of issue is awkward to debug and resolve, but maybe there is something between the versions that changed this mechanism?

Love the library, makes keybinds so much easier.

**Comment by TolikPylypchuk** [comment]
Right, I totally forgot you said that it worked with version 4. I will look at the diff between libuiohook used in version 4 and 5 and see whether I can find anything.

**Comment by TigerHix** [comment]
Can confirm this bug on Windows 11!

**Comment by TolikPylypchuk** [comment]
Hi @leMicin @TigerHix! I may have fixed the issue in the latest build of SharpHook - it's hard to know since the issue itself is hard to reproduce. Could you please download the latest libuiohook binary and replace it directly in the bin folder of your application to verify that it works for you? If the issue is indeed fixed, then I'll release version 5.1.0 soon (hopefully).

You can get the latest libuiohook binary from the [latest GitHub Actions artifact](https://github.com/TolikPylypchuk/SharpHook/actions/runs/6370126154). Download libuiohook-build and inside it windows/<your arch>/bin/uiohook.dll is the file you need.

**Comment by Herdubreid** [comment]
I've come across the same issue with the Ctrl key, just as described by @leMicin (I'm a SharpHook newbie).
What I have found is that if you put a breakpoint on the `KeyTyped` event, then the Mask sticks until you press the key again (I've tried the `uiohook.dll` from 1/10/23 as well).
Without the breakpoint it seems to work fine.
``` csharp
public partial class MainPage : ContentPage
{
	private void _hook_KeyTyped(object sender, KeyboardHookEventArgs e)
	{
		if (e.Data.KeyCode == KeyCode.VcV &&
			(e.RawEvent.Mask & ModifierMask.Ctrl) != ModifierMask.None)
		{
			// TODO
		}
	}

	private readonly TaskPoolGlobalHook _hook;
	public MainPage()
	{
		InitializeComponent();

		_hook = new TaskPoolGlobalHook();
		_hook.KeyTyped += _hook_KeyTyped;

		Task.Run(async () =>
		{
			await _hook.RunAsync();
		});
	}

	protected override void OnDisappearing()
	{
		base.OnDisappearing();
		_hook.Dispose();
	}
}
```

**Comment by TolikPylypchuk** [comment]
Hi @Herdubreid! This is expected behavior actually. If you pause on a breakpoint, then the global hook won't be able to process events (as all threads are stopped) and so when you release the Ctrl key it doesn't know about it, and so the Ctrl mask stays. Not much I can do about it unfortunately, this is just the way Windows handles global hooks.

**Comment by Herdubreid** [comment]
This makes sense @TolikPylypchuk and not a problem as long as you're aware of it.
Great work by the way!

**Comment by TolikPylypchuk** [comment]
I released version 5.1.0 today which should contain the fix for this issue (at least it's not reproducible for me anymore), so this issue can be closed. If it doesn't fix the issue after all, then we can reopen this issue.

**Comment by leMicin** [comment]
@TolikPylypchuk Hi, thanks for checking this out! I couldn't check earlier but I just updated and tried this scenario again. Everything appears to be working now! I am glad you figured it out.

Thank you!
